File: wasm/navmesh_node.ts
var STATIC_ARRAY_BUFFER_STEP: i32 = 8;  // when increase the data values count in static array and overflow it length, then we recreate new bigger array

export class NavmeshNode {
    m_vertices: StaticArray<f32>;
    m_polygon: StaticArray<i32>;
    m_length: i32;  // the number of corners in the polygon
    m_index: i32;  // index = -1 mean, that node is empty
    m_group: i32;
    m_neighbor: StaticArray<i32>;  // manualy increase the buffer, values are indexes of neighborhood polygons
    m_neighbor_count: i32;
    m_center: StaticArray<f32>;
    m_normal: StaticArray<f32>;
    m_vertex_normals: StaticArray<f32>;  // packed per-vertex [x, y, z for the first vertex, then for the second, ...]
    m_portals: Map<i32, StaticArray<f32>>;  // key - node index, value - array of length 6 with vertex coordinates

    constructor(all_vertices: StaticArray<f32>, index: i32, polygon_indexes: StaticArray<i32>) {
        let indexes_len = polygon_indexes.length;

        this.m_length = indexes_len;
        this.m_index = index;

        let vertices = new StaticArray<f32>(3 * indexes_len);
        this.m_vertices = vertices;

        for (let i = 0; i < indexes_len; i++) {
            let k = unchecked(polygon_indexes[i]);
            for (let j = 0; j < 3; j++) {
                unchecked(vertices[3 * i + j] = all_vertices[3 * k + j]);
            }
        }

        this.m_polygon = polygon_indexes;
        this.m_group = -1;
        this.m_neighbor = new StaticArray<i32>(0);
        this.m_neighbor_count = 0;

        //calculate polygon center
        this.m_center = new StaticArray<f32>(3);
        this.m_normal = new StaticArray<f32>(3);
        this.m_vertex_normals = new StaticArray<f32>(3 * indexes_len);
        this.m_portals = new Map<i32, StaticArray<f32>>();
        // init zero values. Unnecessary in AS
        // for (let i = 0; i < 3; i++){
        //     this.m_center[i] = 0.0;
        //     this.m_normal[i] = 0.0;
        // }

        let cx = unchecked(this.m_center[0]);
        let cy = unchecked(this.m_center[1]);
        let cz = unchecked(this.m_center[2]);

        for (let i = 0; i < indexes_len; i++) {
            cx += unchecked(vertices[3 * i + 0]);
            cy += unchecked(vertices[3 * i + 1]);
            cz += unchecked(vertices[3 * i + 2]);
        }
        if (indexes_len > 0) {
            let invLen: f32 = 1.0 / <f32>indexes_len;
            cx *= invLen;
            cy *= invLen;
            cz *= invLen;
        }

        unchecked(this.m_center[0] = cx);
        unchecked(this.m_center[1] = cy);
        unchecked(this.m_center[2] = cz);

        //after center we can calculate average normal of the polygon
        for (let i = 0; i < indexes_len; i++) {
            let j = i + 1;
            if (j == indexes_len) {
                j = 0;
            }
            let c_x = this._cross_x(
                unchecked(vertices[3 * i + 1]) - cy,
                unchecked(vertices[3 * i + 2]) - cz,
                unchecked(vertices[3 * j + 1]) - cy,
                unchecked(vertices[3 * j + 2]) - cz
            );
            let c_y = this._cross_y(
                unchecked(vertices[3 * i + 0]) - cx,
                unchecked(vertices[3 * i + 2]) - cz,
                unchecked(vertices[3 * j + 0]) - cx,
                unchecked(vertices[3 * j + 2]) - cz
            );
            let c_z = this._cross_z(
                unchecked(vertices[3 * i + 0]) - cx,
                unchecked(vertices[3 * i + 1]) - cy,
                unchecked(vertices[3 * j + 0]) - cx,
                unchecked(vertices[3 * j + 1]) - cy
            );
            unchecked(this.m_normal[0] += c_x);
            unchecked(this.m_normal[1] += c_y);
            unchecked(this.m_normal[2] += c_z);
        }
        //normalize the normal
        if (indexes_len > 0) {
            let nx = unchecked(this.m_normal[0]);
            let ny = unchecked(this.m_normal[1]);
            let nz = unchecked(this.m_normal[2]);
            let invLen: f32 = 1.0 / Mathf.sqrt(nx * nx + ny * ny + nz * nz);
            unchecked(this.m_normal[0] = nx * invLen);
            unchecked(this.m_normal[1] = ny * invLen);
            unchecked(this.m_normal[2] = nz * invLen);
        }

        let vertex_normals = this.m_vertex_normals;

        //vertex normals
        for (let i = 0, len = 3 * indexes_len; i < len; i++) {
            unchecked(vertex_normals[i] = 0.0);
        }

        for (let i = 0; i < indexes_len; i++) {
            let j = i + 1;
            if (j == indexes_len) {
                j = 0;
            }
            let k = j + 1;
            if (k == indexes_len) {
                k = 0;
            }
            let v_x = this._cross_x(
                unchecked(vertices[3 * j + 1] - vertices[3 * i + 1]),
                unchecked(vertices[3 * j + 2] - vertices[3 * i + 2]),
                unchecked(vertices[3 * k + 1] - vertices[3 * i + 1]),
                unchecked(vertices[3 * k + 2] - vertices[3 * i + 2])
            );
            let v_y = this._cross_y(
                unchecked(vertices[3 * j + 0] - vertices[3 * i + 0]),
                unchecked(vertices[3 * j + 2] - vertices[3 * i + 2]),
                unchecked(vertices[3 * k + 0] - vertices[3 * i + 0]),
                unchecked(vertices[3 * k + 2] - vertices[3 * i + 2])
            );
            let v_z = this._cross_z(
                unchecked(vertices[3 * j + 0] - vertices[3 * i + 0]),
                unchecked(vertices[3 * j + 1] - vertices[3 * i + 1]),
                unchecked(vertices[3 * k + 0] - vertices[3 * i + 0]),
                unchecked(vertices[3 * k + 1] - vertices[3 * i + 1])
            );

            let inv_d: f32 = 1.0 / Mathf.sqrt(v_x * v_x + v_y * v_y + v_z * v_z);
            unchecked(vertex_normals[3 * i + 0] = v_x * inv_d);
            unchecked(vertex_normals[3 * i + 1] = v_y * inv_d);
            unchecked(vertex_normals[3 * i + 2] = v_z * inv_d);
        }
    }

    add_neighbor(node_index: i32, v0_x: f32, v0_y: f32, v0_z: f32, v1_x: f32, v1_y: f32, v1_z: f32): void {
        if (!this.m_portals.has(node_index)) {
            var new_portal = new StaticArray<f32>(6);
            unchecked(new_portal[0] = v0_x);
            unchecked(new_portal[1] = v0_y);
            unchecked(new_portal[2] = v0_z);
            unchecked(new_portal[3] = v1_x);
            unchecked(new_portal[4] = v1_y);
            unchecked(new_portal[5] = v1_z);

            this.m_portals.set(node_index, new_portal);

            let neighbor = this.m_neighbor;
            let neighbor_len = this.m_neighbor.length;
            if (this.m_neighbor_count == neighbor_len) {
                //create new neighbor buffer
                var new_neighbor = new StaticArray<i32>(neighbor_len + STATIC_ARRAY_BUFFER_STEP);
                //copy values to them
                for (let i = 0; i < neighbor_len; i++) {
                    unchecked(new_neighbor[i] = neighbor[i]);
                }
                //rewrite neigboor link
                this.m_neighbor = new_neighbor;
            }

            this.m_neighbor_count++;
            unchecked(neighbor[this.m_neighbor_count - 1] = node_index);
        }
    }

    @inline
    get_portal(node_index: i32): StaticArray<f32> {
        let portals = this.m_portals;
        if (portals.has(node_index)) {
            return portals.get(node_index);
        } else {
            return new StaticArray<f32>(0);
        }
    }

    get_neighbord(): StaticArray<i32> {
        //for return we create new array of actual size
        //in main program this method called only on constructor, so, it does not use GC many times
        let neighbor = this.m_neighbor;
        let neighbor_count = this.m_neighbor_count;
        let to_return = new StaticArray<i32>(neighbor_count);
        for (let i = 0; i < neighbor_count; i++) {
            unchecked(to_return[i] = neighbor[i]);
        }
        return to_return;
    }

    @inline
    get_index(): i32 {
        return this.m_index;
    }

    @inline
    get_group(): i32 {
        return this.m_group;
    }

    @inline
    get_vertex_indexes(): StaticArray<i32> {
        return this.m_polygon;
    }

    @inline
    get_vertex_coordinates(): StaticArray<f32> {
        return this.m_vertices;
    }

    @inline
    get_normal(): StaticArray<f32> {
        return this.m_normal;
    }

    @inline
    get_center(): StaticArray<f32> {
        return this.m_center;
    }

    set_group(group_index: i32, group_array: Array<i32>, all_nodes: StaticArray<NavmeshNode>): void {
        if (this.m_group == -1) {
            this.m_group = group_index;
            let neighbor = this.m_neighbor;
            group_array.push(this.m_index);  // <-- recreate array here
            for (let i = 0, len = this.m_neighbor_count; i < len; i++){
                unchecked(all_nodes[neighbor[i]]).set_group(group_index, group_array, all_nodes);
            }
        }
    }

    is_point_inside(p_x: f32, p_y: f32, p_z: f32): bool {
        let vertices = this.m_vertices;
        let normals = this.m_vertex_normals;
        for (let i = 0, len = this.m_length; i < len; i++) {
            let j = i + 1;
            if (j == len) {
                j = 0;
            }
            let v_x = this._cross_x(
                unchecked(vertices[3 * j + 1] - vertices[3 * i + 1]),
                unchecked(vertices[3 * j + 2] - vertices[3 * i + 2]),
                p_y - unchecked(vertices[3 * i + 1]),
                p_z - unchecked(vertices[3 * i + 2])
            );
            let v_y = this._cross_y(
                unchecked(vertices[3 * j + 0] - vertices[3 * i + 0]),
                unchecked(vertices[3 * j + 2] - vertices[3 * i + 2]),
                p_x - unchecked(vertices[3 * i + 0]),
                p_z - unchecked(vertices[3 * i + 2])
            );
            let v_z = this._cross_z(
                unchecked(vertices[3 * j + 0] - vertices[3 * i + 0]),
                unchecked(vertices[3 * j + 1] - vertices[3 * i + 1]),
                p_x - unchecked(vertices[3 * i + 0]),
                p_y - unchecked(vertices[3 * i + 1])
            );
            let d = (
                v_x * unchecked(normals[3 * i + 0]) +
                v_y * unchecked(normals[3 * i + 1]) +
                v_z * unchecked(normals[3 * i + 2])
            );
            if (d < 0) {
                return false;
            }
        }
        return true;
    }

    @inline
    _cross_x(a_y: f32, a_z: f32, b_y: f32, b_z: f32): f32 {
        return a_y * b_z - a_z * b_y;
    }

    @inline
    _cross_y(a_x: f32, a_z: f32, b_x: f32, b_z: f32): f32 {
        return a_z * b_x - a_x * b_z;
    }

    @inline
    _cross_z(a_x: f32, a_y: f32, b_x: f32, b_y: f32): f32 {
        return a_x * b_y - a_y * b_x;
    }

    @inline
    _cross(a_x: f32, a_y: f32, a_z: f32, b_x: f32, b_y: f32, b_z: f32): StaticArray<f32> {
        let to_return = new StaticArray<f32>(3);
        unchecked(to_return[0] = a_y * b_z - a_z * b_y);
        unchecked(to_return[1] = a_z * b_x - a_x * b_z);
        unchecked(to_return[2] = a_x * b_y - a_y * b_x);
        return to_return;
    }

    _portals_to_string(): string {
        var to_return = "{";
        let keys = this.m_portals.keys();
        for (let i = 0, len = keys.length; i < len; i++) {
            let key = unchecked(keys[i]);
            to_return += (
                key.toString() + ": " +
                this.m_portals.get(key).toString()
            );
            if (i < len - 1) {
                to_return += ", ";
            }
        }
        return to_return + "}";
    }

    to_string(): string {
        return "<node " + this.m_index.toString() +
               ", vertices: " + this.m_polygon.toString() +
               ", positions: " + this.m_vertices.toString() +
               ", center: " + this.m_center.toString() +
               ", normal: " + this.m_normal.toString() +
               ", vertex normals: " + this.m_vertex_normals.toString() +
               ", group: " + this.m_group.toString() +
               ", neighbor: " + this.m_neighbor.toString() + " (" + this.m_neighbor_count.toString() + ")" +
               ", portals: " + this._portals_to_string() +
               ">";
    }

    toString(): string {
        return this.to_string();
    }
}

File: python/pathfinder/navmesh/navmesh_node.py
import math
from typing import List, Tuple, Optional, Dict


class NavmeshNode:
    '''Class for representing one node in the navigation mesh. Each node is a polygon with some additional data:
        -) group index - the index of the connected component of polygons
        -) polygon index
        -) array of indexes of incident polygons
        -) indexes of polygon vertices
        -) coordinates of the polygon center
        -) portals (this is a map, where key - index of the neighbor node, value - the pair of 3-tuples ((v1), (v2)), where v1 and v2 coordinates of the edge vertices )
                    ordered in orientetion, induced from polygon orientation

        polygons are convex, it's important for calculation it normal

        this class is for internal use only
    '''
    def __init__(self, all_vertices: List[Tuple[float, float, float]], index: int, polygon_indexes: List[int]):
        '''Pass to the constructor all vertices of the navigation mesh
        '''
        self._vertices: List[Tuple[float, float, float]] = [all_vertices[i] for i in polygon_indexes]
        self._polygon: List[int] = polygon_indexes  # indexes of polygon vertices
        self._group: int = -1
        self._index: int = index
        self._neighbors: List[int] = []
        self._center: Tuple[float, float, float] = self._calc_center(self._vertices)
        self._normal: Tuple[float, float, float] = self._calc_average_normal(self._center, self._vertices)
        self._portals: Dict[int, Tuple[Tuple[float, float, float], Tuple[float, float, float]]] = {}
        self._vertex_normals: List[Tuple[float, float, float]] = self._calc_vertex_normals(self._vertices)

    def add_neighbor(self, node_index: int, vertex_0: Tuple[float, float, float], vertex_1: Tuple[float, float, float]):
        '''add neighbor polygon and portal
        '''
        if node_index not in self._neighbors:
            self._neighbors.append(node_index)
            self._portals[node_index] = (vertex_0, vertex_1)
        else:
            print("[Something wrong] Try to add neighbor node " + str(node_index) + " to the " + str(self._index) + ", but this node already exists")

    def get_polygon(self) -> List[int]:
        return self._polygon

    def get_portal(self, node_index: int) -> Tuple[Tuple[float, float, float], Tuple[float, float, float]]:
        '''return pair of 3-tuples with coordinates of the edge, which separate current node from node with index node_index
        '''
        return self._portals[node_index]

    def get_neighbord(self) -> List[int]:
        return self._neighbors

    def get_index(self) -> int:
        return self._index

    def get_group(self) -> int:
        return self._group

    def get_vertex_indexes(self) -> List[int]:
        return self._polygon

    def get_vertex_coordinates(self) -> List[Tuple[float, float, float]]:
        return self._vertices

    def get_normal(self) -> Tuple[float, float, float]:
        return self._normal

    def get_center(self) -> Tuple[float, float, float]:
        return self._center

    def set_group(self, group_index: int, group_array: List[int], all_nodes):  # all_nodes: List[NavmeshNode]
        if self._group == -1:
            self._group = group_index
            group_array.append(self._index)
            for n in self._neighbors:
                all_nodes[n].set_group(group_index, group_array, all_nodes)

    def is_point_inside(self, point: Tuple[float, float, float]) -> bool:
        '''return true, if the point inside the polygon
        '''
        size: int = len(self._vertices)
        for i in range(size):
            u: Tuple[float, float, float] = self._vertices[i]
            v: Tuple[float, float, float] = self._vertices[self._modulize(i + 1, size)]
            # we should calculate cross product [uv, up]
            # if dot-product with normal in the vertex < 0, then point outside the polygon
            vector: Tuple[float, float, float] = self._cross(v[0] - u[0], v[1] - u[1], v[2] - u[2], point[0] - u[0], point[1] - u[1], point[2] - u[2])
            d: float = vector[0] * self._vertex_normals[i][0] + vector[1] * self._vertex_normals[i][1] + vector[2] * self._vertex_normals[i][2]
            if d < 0.0:
                return False
        return True

    def _modulize(self, value: int, n: int) -> int:
        if value < n:
            return value
        else:
            return value - n

    def _calc_vertex_normals(self, vertices: List[Tuple[float, float, float]]) -> List[Tuple[float, float, float]]:
        '''for each vertex u (and next v, w) calculate normalized [uv, uw]
        '''
        size: int = len(vertices)
        to_return: List[Tuple[float, float, float]] = []
        for i in range(size):
            u: Tuple[float, float, float] = vertices[i]
            v: Tuple[float, float, float] = vertices[self._modulize(i + 1, size)]
            w: Tuple[float, float, float] = vertices[self._modulize(i + 2, size)]
            vector: Tuple[float, float, float] = self._cross(v[0] - u[0], v[1] - u[1], v[2] - u[2], w[0] - u[0], w[1] - u[1], w[2] - u[2])
            d: float = math.sqrt(vector[0]**2 + vector[1]**2 + vector[2]**2)
            to_return.append((vector[0] / d, vector[1] / d, vector[2] / d))
        return to_return

    def _calc_center(self, array: List[Tuple[float, float, float]]) -> Tuple[float, float, float]:
        '''Find center of the array
        '''
        s = [0.0] * 3
        for a in array:
            for i in range(len(a)):
                s[i] += a[i]
        for i in range(len(s)):
            s[i] = s[i] / len(array)
        return (s[0], s[1], s[2])

    def _cross(self, a_x: float, a_y: float, a_z: float, b_x: float, b_y: float, b_z: float) -> Tuple[float, float, float]:
        return (a_y*b_z - a_z*b_y, a_z*b_x - a_x*b_z, a_x*b_y - a_y*b_x)

    def _calc_average_normal(self, center: Tuple[float, float, float], points: List[Tuple[float, float, float]]) -> Tuple[float, float, float]:
        normal: List[float] = [0.0, 0.0, 0.0]
        for p_index in range(len(points)):
            p0: Tuple[float, float, float] = points[p_index]
            p1: Tuple[float, float, float] = points[p_index + 1 if p_index < len(points) - 1 else 0]
            c: Tuple[float, float, float] = self._cross(p0[0] - center[0], p0[1] - center[1], p0[2] - center[2], p1[0] - center[0], p1[1] - center[1], p1[2] - center[2])
            normal[0] += c[0]
            normal[1] += c[1]
            normal[2] += c[2]
        d: float = math.sqrt(normal[0]**2 + normal[1]**2 + normal[2]**2)
        return (normal[0] / d, normal[1] / d, normal[2] / d)

    def __repr__(self):
        return "<node " + str(self._index) +\
               ", vertices: " + str(self._polygon) +\
               ", normal: " + str(self._normal) +\
               ", local normals: " + str(self._vertex_normals) +\
               ", neigh: " + str(self._neighbors) +\
               ", component: " + str(self._group) +\
               ", portals: " + str(self._portals) +\
               ">"
